source/opt/instruction.cpp

```diff
@@ -701,5 +701,27 @@ bool Instruction::IsScalarizable() const {
   return false;
 }
 
+bool Instruction::IsOpcodeSafeToDelete() const {
+  if (context()->IsCombinatorInstruction(this)) {
+    return true;
+  }
+
+  switch (opcode()) {
+    case SpvOpDPdx:
+    case SpvOpDPdy:
+    case SpvOpFwidth:
+    case SpvOpDPdxFine:
+    case SpvOpDPdyFine:
+    case SpvOpFwidthFine:
+    case SpvOpDPdxCoarse:
+    case SpvOpDPdyCoarse:
+    case SpvOpFwidthCoarse:
+    case SpvOpImageQueryLod:
```

Comment:
Shouldn't it be more consistent to move this ImageQuery op to AddCombinatorsForCapability?

Reply:
No because it is not a combinator.  In the description of the opcode:

> This instruction is only valid in the Fragment Execution Model. **In addition, it consumes an implicit derivative that can be affected by code motion.**

Reply:
Ah.  Thanks.